R/percent.R

```diff
@@ -0,0 +1,194 @@
+#' Percentages
+#'
+#' @description
+#' `percent` is a lightweight S3 class allowing for pretty
+#' printing of proportions as percentages.
+#'
+#' @param x [numeric] vector of proportions.
+#'
+#' @returns
+#' A class of object `percent`.
+#'
+#' @details
+#' By default all percentages are formatted to 2 decimal places which can be
+#' overwritten using `format()`. It's worth noting that the digits argument in
+#' `format.percent` uses decimal rounding instead of significant digit rounding.
+#'
+#' @examples
+#' library(phsmethods)
+#'
+#' # Convert proportions to percentages
+#' as_percent(seq(0, 1, 0.1))
+#'
+#' # You can use round() as usual
+#' p <- as_percent(15.56 / 100)
+#' round(p)
+#' round(p, digits = 1)
+#'
+#' p2 <- as_percent(0.0005)
+#' signif(p2, 2)
+#' floor(p2)
+#' ceiling(p2)
+#'
+#' # We can do basic math operations as usual
+#'
+#' # Order of operations matters
+#' 10 * as_percent(c(0, 0.5, 2))
+#' as_percent(c(0, 0.5, 2)) * 10
+#'
+#' as_percent(0.1) + as_percent(0.2)
+#'
+#' # Formatting options
+#' format(as_percent(2.674 / 100), digits = 2, symbol = " (%)")
+#' # Prints nicely in data frames (and tibbles)
+#' library(dplyr)
+#' starwars %>%
+#'   count(eye_color) %>%
+#'   mutate(perc = as_percent(n/sum(n))) %>%
+#'   arrange(desc(perc))  %>% # We can do numeric sorting with percent vectors
+#'   mutate(perc_rounded = round(perc))
+#' @export
+#' @rdname percent
+as_percent <- function(x){
+  if (!is.numeric(x)){
+    stop("x must be a numeric vector of proportions")
+  }
+  new_percent(x)
+}
+new_percent <- function(x){
+  class(x) <- "percent"
+  x
+}
+round_half_up <- function(x, digits = 0){
+  if (is.null(digits) || (length(digits) == 1 && digits == Inf)){
+    return(x)
+  }
+  trunc(
+    abs(x) * 10^digits + 0.5 +
+      sqrt(.Machine$double.eps)
+  ) /
+    10^digits * sign(x)
+}
+signif_half_up <- function(x, digits = 6){
+  if (is.null(digits) || (length(digits) == 1 && digits == Inf)){
+    return(x)
+  }
+  round_half_up(x, digits - ceiling(log10(abs(x))))
+}
+
+#' @export
+as.character.percent <- function(x, digits = 2, ...){
+  if (length(x) == 0){
+    character()
+  } else {
+    paste0(unclass(round(x, digits) * 100), "%")
+  }
+}
+
+#' @export
+format.percent <- function(x, symbol = "%", trim = TRUE,
+                           digits = 2,
+                           ...){
+  if (length(x) == 0){
+    out <- character()
+  } else {
+    out <- paste0(format(unclass(round(x, digits) * 100), trim = trim, digits = NULL, ...),
+                  symbol)
```

Comment:
Not sure if I'm right here, but wouldn't your code use the base round methodology?

```suggestion
    out <- paste0(round(x, digits = digits), trim = trim, digits = NULL, ...),
                  symbol)
```

Reply:
No I don't think so because `round()` is nested inside the `unclass()` call.